docs/design/labels.md

```diff
@@ -0,0 +1,239 @@
+③ Tell the pipeline which element from the json to send to the next stage.  
+
+```yaml
+    - regex:
+        expr: '.*level=(?P<level>[a-zA-Z]+).*'  ①
+        labels:
+          level:                                ②
+```
+
+① Define the Go RE2 regex, making sure to use a named capture group.  
+② Extract labels using the named capture group names.  
+
+Notice there was not an output section defined here, omitting the output key should instruct the parser to return the incoming log message to the next stage with no changes.
+
+```yaml
+    - regex:
+        expr: '.*msg=(?P<message>[a-zA-Z]+).*'
+        output: message                          ①
+```
+
+① Send the log message as the output to the last stage in the pipeline, this will be what you want Loki to store as the log message.
+
+There is an alternative configuration that could be used here to accomplish the same result:
+
+```yaml
+scrape_configs:
+- job_name: system
+  entry_parsers:
+    - json: 
+        timestamp:
+          source: time
+          format: FIXME
+        labels:
+          stream:
+        output: log
+    - regex:
+        expr: '.*level=(?P<level>[a-zA-Z]+).*msg=(?P<message>[a-zA-Z]+).*'
+        labels:
+          level:                                                             ①
+          log:
+            source: message                                                  ②
+        output: message
+```
+
+① Similar to the json parser, if your log label matches the regex named group, you need only specify the label name as a yaml key  
+② If you had a use case for specifying a different label name from the regex group name you can optionally provide the `source` key with the value matching the named capture group.
+
+You can define a more complicated regular expression with multiple capture groups to extract many labels and/or the output log message in one entry parser.  This has the advantage of being more performant, however, the regular expression will also get much more complicated.
+
+Please also note the regex for `message` is incomplete and would do a terrible job of matching any standard log message which might contain spaces or non alpha characters.
+
+### Concerns
+* Debugging, especially if a pipeline stage is mutating the log entry. 
+* Clashing labels and how to handle this (two stages try to set the same label) 
+* Performance vs ease of writing/use, if every label is extracted one at a time and there are a lot of labels and a long line, it would force reading the line many times, however contrast this to a really long complicated regex which only has to read the line once but is difficult to write and/or change and maintain
```

Comment:
What about multi-line logging?  For example, both docker json-file logs, and crio file logs, are multi-line by default - with docker, if the line length is greater than 16k, you will get partial lines in the log file that must be reconstructed before further parsing - with cri-o, you will almost certainly get partial lines due to the way it flushes the buffers to log files.

But perhaps multi-line logging is not a concern of loki, and it is up to the user of loki to figure out what they want to do with multi-line logs?  Not to mention multi-line stacktraces, etc.